### PR TITLE
removing _pixiContainer and _pixiTexture from PhaserProjectile.ts

### PR DESCRIPTION
**Pull request by Grenagar**
just my ideas and starting steps on removing PIXI from projectiles

### CODE REVIEW

File: ts/src/renderer/phaser/classes/PhaserProjectile.ts [hunk 1]
```diff
@@ -22,27 +24,32 @@ class PhaserProjectile extends Phaser.GameObjects.Container {
...
+		this.x = projectile._translate.x;
+		this.y = projectile._translate.y;
+		//this.x = container.x;
+		//this.y = container.y;
 
 		const sprite = this.sprite;
-		sprite.rotation = texture.rotation;
-		sprite.setScale(texture.scale.x, texture.scale.y);
+		sprite.rotation = projectile._rotate.z; //texture.rotation;
+		//sprite.setScale (projectile._scale.x, projectile._scale.y); //texture.scale.x, texture.scale.y);
```
Comment (pavle-goloskokovic) [inline ts/src/renderer/phaser/classes/PhaserProjectile.ts:53]: can a projectile change scale compared to the initial set at creation time? if so, we should update it accordingly here

File: ts/src/renderer/phaser/scenes/GameScene.ts [hunk 2]
```diff
@@ -9,7 +9,7 @@ class GameScene extends Phaser.Scene {
 		// phaser canvas adjustments
 		const canvas = this.game.canvas;
 		canvas.style.position = 'fixed';
-		canvas.style.opacity = '0.5';
+		//canvas.style.opacity = '0.5';
```
Comment (pavle-goloskokovic) [inline ts/src/renderer/phaser/scenes/GameScene.ts:12]: unrelated change, should be reverted

File: ts/src/renderer/phaser/classes/PhaserProjectile.ts [hunk 3]
```diff
@@ -22,27 +24,32 @@ class PhaserProjectile extends Phaser.GameObjects.Container {
 				console.log('PhaserProjectile play-animation', `${key}/${animationId}`);  // TODO remove
 				sprite.play(`${key}/${animationId}`);
 			});
+
+		//console.log('projectile', projectile._scale, projectile._pixiTexture.scale, projectile._stats.currentBody.width)
```
Comment (pavle-goloskokovic) [inline ts/src/renderer/phaser/classes/PhaserProjectile.ts:28]: cleanup later

File: ts/src/renderer/phaser/classes/PhaserProjectile.ts [hunk 4]
```diff
@@ -22,27 +24,32 @@ class PhaserProjectile extends Phaser.GameObjects.Container {
...
 	update (/*time: number, delta: number*/): void {
 
 		const projectile = this.projectile;
-		const container = projectile._pixiContainer;
-		const texture = projectile._pixiTexture;
+		//const container = projectile._pixiContainer;
+		//const texture = projectile._pixiTexture;
 
-		if (projectile._destroyed || container._destroyed) {
+		if (!projectile._alive/*projectile._destroyed || container._destroyed*/) {
```
Comment (pavle-goloskokovic) [inline ts/src/renderer/phaser/classes/PhaserProjectile.ts:37]: `_alive` seems like a good alternative to `__destroyed`

File: ts/src/renderer/phaser/classes/PhaserProjectile.ts [hunk 5]
```diff
@@ -12,6 +12,8 @@ class PhaserProjectile extends Phaser.GameObjects.Container {
 		const key = `projectile/${projectile._stats.type}`;
 
 		const sprite = this.sprite = scene.add.sprite(0, 0, key);
+		sprite.displayWidth = projectile._stats.currentBody.width;
+		sprite.displayHeight = projectile._stats.currentBody.height;
```
Comment (pavle-goloskokovic) [inline ts/src/renderer/phaser/classes/PhaserProjectile.ts:16]: these look like great alternative to reading pixi object properties. 
are `currentBody` and its `width` and `height` always defined when creating a projectile? 

example: https://github.com/moddio/taro/blob/57ddd8db7c5e352ebb3ccce94492002e3a7c5a34/src/pixi/IgePixiAnimation.js#L59-L60

Comment (sloont) [reply ts/src/renderer/phaser/classes/PhaserProjectile.ts]: Can you reference a time where `.currentBody` doesn't exist? These are the types of issues I was looking out for when we were doing Crash.

Comment (pavle-goloskokovic) [reply ts/src/renderer/phaser/classes/PhaserProjectile.ts:17]: not specifically, but existence of conditions in above code does indicate that there are cases when it doesn't exist.

File: src/gameClasses/Projectile.js [hunk 6]
```diff
@@ -86,17 +83,11 @@ var Projectile = IgeEntityPhysics.extend({
 			ige.server.totalProjectilesCreated++;
 		} else if (ige.isClient) {
 			if (currentState) {
-				var defaultAnimation = this._stats.animations[currentState.animation];
-				this.createPixiTexture(defaultAnimation && defaultAnimation.frames[0] - 1, data.defaultData);
+				/* TEMPORARY */
+				//TODO - add new logic for trackEntityById and updating entities with out PIXI in EntityManager
+				ige.pixi.trackEntityById[this.entityId] = undefined;
```
Comment (pavle-goloskokovic) [inline src/gameClasses/Projectile.js:88]: this should probably be set to `this` and without the condition.

Also, I've noticed animations not working anymore (explosion in default game).

File: ts/types/IgeEntity.d.ts [hunk 7]
```diff
@@ -1,11 +1,14 @@
...
 
+	_alive: boolean;
 	_destroyed: boolean;
 	_deathTime: number;
 	_category: string; // TODO more specific values
 	_translate: IgePoint3d;
 	_rotate: IgePoint3d;
+	_scale: IgePoint3d;
 	_stats: EntityStats;
+	_bounds2d: IgePoint3d;
```
Comment (pavle-goloskokovic) [inline ts/types/IgeEntity.d.ts:11]: correct type is `IgePoint2d`

Comment (sloont) [reply ts/types/IgeEntity.d.ts:11]: there are some really unfortunate places in the source that I have found in the past where IgePoint3d is used and the rotation value is passed as the z coord. If we can remove these as we find them that's awesome, but we'll have to make note of all these.

File: ts/src/renderer/phaser/classes/PhaserProjectile.ts (hunk 4 above)
Comment (sloont) [reply ts/src/renderer/phaser/classes/PhaserProjectile.ts]: i think the intention is `._destroys` indicates it is marked for clean-up

Comment (pavle-goloskokovic) [reply ts/src/renderer/phaser/classes/PhaserProjectile.ts]: good point, might be smart to keep track of `_destroyed` in ige entity, same as is now for pixi object, and query that instead

File: ts/src/renderer/phaser/classes/PhaserProjectile.ts (hunk 5 above)
Comment (pavle-goloskokovic) [reply ts/src/renderer/phaser/classes/PhaserProjectile.ts:17]: after doing some work with unit scaling I think the correct data is stored in `_bounds2d` as it gets updated when changing entity's width and height consistently

File: engine/core/IgeEntity.js [hunk 8]
```diff
@@ -640,7 +640,8 @@ var IgeEntity = IgeObject.extend({
 
 		return this._backgroundPattern;
 	},
-	createTexture: function (defaultSprite = 0, defaultData) {
+	addToRenderer: function (defaultSprite = 0, defaultData) {
```
Comment (sloont) [inline engine/core/IgeEntity.js:643]: I think this is a great change, but normally it would be outside of the scope of this PR

File: ts/types/EntityStats.d.ts [hunk 9]
```diff
@@ -3,6 +3,8 @@ declare interface EntityStats {
 	name: string;
 	currentBody: {
 		jointType: string;
+		width: number;
```
Comment (sloont) [inline ts/types/EntityStats.d.ts:6]: didn't we decide to use bounds2d?

File: engine/core/IgeEntity.js (hunk 8 above)
Comment (Grenagar) [reply engine/core/IgeEntity.js:643]: createTexture is just confusing name now, so i renamed

File: ts/types/EntityStats.d.ts (hunk 9 above)
Comment (Grenagar) [reply ts/types/EntityStats.d.ts:6]: we use bounds2d now, yes, you think we should remove width from types?

File: engine/core/IgeEntity.js (hunk 8 above)
Comment (pavle-goloskokovic) [reply engine/core/IgeEntity.js:643]: I'm ok with this renaming here, the line is blurred to say with certainty that it is outside of the scope, let's not overthink this one and just keep it in here ;)

Comment (sloont) [reply engine/core/IgeEntity.js:643]: i agree with keeping it in here. the work is done :D

File: ts/types/EntityStats.d.ts (hunk 9 above)
Comment (pavle-goloskokovic) [reply ts/types/EntityStats.d.ts:6]: if we put in work to figure out types of data and objects in taro, I would keep it in as it might be useful later, no need to dismiss the effort even if it is not used currently

File: ts/src/renderer/phaser/classes/PhaserProjectile.ts [hunk 10]
```diff
@@ -12,37 +15,44 @@ class PhaserProjectile extends Phaser.GameObjects.Container {
...
 		scene.add.existing(this);
 		scene.events.on('update', this.update, this);
 
+		this.transformListener = projectile.on('transform', (info) => {
+			this.x = info.x;
+			this.y = info.y;
+			this.sprite.rotation = info.rotation;
+		});
+
+		this.scaleListener = projectile.on('scale', (info) => {
```
Comment (pavle-goloskokovic) [inline ts/src/renderer/phaser/classes/PhaserProjectile.ts:36]: same as for `transform`

File: ts/src/renderer/phaser/classes/PhaserProjectile.ts [hunk 11]
```diff
@@ -21,31 +24,32 @@ class PhaserProjectile extends Phaser.GameObjects.Container {
...
 			});
-	}
-
-	update (/*time: number, delta: number*/): void {
-
-		const projectile = this.projectile;
 
-		if (!projectile._alive) {
+		this.destroyListener = projectile.on('destroy', () => {
 			projectile.off('play-animation', this.playAnimationListener);
```
Comment (sloont) [inline ts/src/renderer/phaser/classes/PhaserProjectile.ts:48]: should we remove the other 3 event listeners in the same manner?

File: ts/src/renderer/phaser/classes/PhaserProjectile.ts [hunk 12]
```diff
@@ -12,37 +15,44 @@ class PhaserProjectile extends Phaser.GameObjects.Container {
 		const key = `projectile/${projectile._stats.type}`;
 
 		const sprite = this.sprite = scene.add.sprite(0, 0, key);
+		const translate = projectile._translate;
+		const bounds = projectile._bounds2d;
+		this.x = translate.x;
```
Comment (pavle-goloskokovic) [inline ts/src/renderer/phaser/classes/PhaserProjectile.ts:20]: we can use `setPosition` here, similar to how it is done below with `setDisplaySize`

File: ts/src/renderer/phaser/classes/PhaserProjectile.ts (hunk 11 above)
Comment (pavle-goloskokovic) [reply ts/src/renderer/phaser/classes/PhaserProjectile.ts]: most definitely, good catch

File: ts/src/renderer/phaser/classes/PhaserProjectile.ts [hunk 13]
```diff
@@ -12,37 +15,41 @@ class PhaserProjectile extends Phaser.GameObjects.Container {
...
+		sprite.setDisplaySize(bounds.x, bounds.y);
+		sprite.rotation = projectile._rotate.z;
+
 		this.add(sprite);
 
 		scene.add.existing(this);
 		scene.events.on('update', this.update, this);
 
+		this.transformListener = projectile.on('transform', (info) => {
+			this.x = info.x;
```
Comment (pavle-goloskokovic) [inline ts/src/renderer/phaser/classes/PhaserProjectile.ts:31]: can use `setPosition` here

File: ts/src/renderer/phaser/classes/PhaserProjectile.ts (hunk 11 above)
Comment (pavle-goloskokovic) [reply ts/src/renderer/phaser/classes/PhaserProjectile.ts]: yes, for the `update` listener

**Review by pavle-goloskokovic — CHANGES_REQUESTED** [review]

File: ts/src/renderer/phaser/classes/PhaserProjectile.ts [hunk 14]
```diff
@@ -45,10 +44,13 @@ class PhaserProjectile extends Phaser.GameObjects.Container {
...
 
 		this.destroyListener = projectile.on('destroy', () => {
-			projectile.off('play-animation', this.playAnimationListener);
+			projectile.off('transform', this.playAnimationListener);
 			this.transformListener = null;
+			projectile.off('scale', this.playAnimationListener);
 			this.scaleListener = null;
+			projectile.off('play-animation', this.playAnimationListener);
 			this.playAnimationListener = null;
+			projectile.off('destroy', this.playAnimationListener);
```
Comment (sloont) [inline ts/src/renderer/phaser/classes/PhaserProjectile.ts:53]: these all say `this.playAnimationListener` :P

Comment (pavle-goloskokovic) [reply ts/src/renderer/phaser/classes/PhaserProjectile.ts:53]: :/

File: ts/src/renderer/phaser/classes/PhaserProjectile.ts (hunk 11 above)
Comment (sloont) [reply ts/src/renderer/phaser/classes/PhaserProjectile.ts]: for destroying these Phaser entities, would it be better to make this a method of PhaserEntity when we introduce it, and just generalize the process by iterating the Entity._eventListeners array and calling `.off()` on them all?

Comment (sloont) [reply ts/src/renderer/phaser/classes/PhaserProjectile.ts]: obviously that would not be part of this PR

Comment (Grenagar) [reply ts/src/renderer/phaser/classes/PhaserProjectile.ts]: yes, good idea, now when we are working on phaser projectile we will gather ideas for phaserEntity PR

File: ts/src/renderer/phaser/classes/PhaserProjectile.ts (hunk 12 above)
Comment (sloont) [reply ts/src/renderer/phaser/classes/PhaserProjectile.ts]: which method is this `setPositon()`? The only reference I'm finding is a box2d method

Comment (sloont) [reply ts/src/renderer/phaser/classes/PhaserProjectile.ts]: nvm I found it. As expected there was no reference in the repo because it belongs to the Phaser source. Cool.

**Review by sloont — APPROVED** [review]
I'm satisfied :D

File: ts/src/renderer/phaser/classes/PhaserProjectile.ts [hunk 15]
```diff
@@ -12,37 +15,42 @@ class PhaserProjectile extends Phaser.GameObjects.Container {
...
 
 		scene.add.existing(this);
 		scene.events.on('update', this.update, this);
 
+		this.transformListener = projectile.on('transform', (data: EntityTransform) => {
+			this.setPosition(data.x, data.y);
+			this.sprite.rotation = data.rotation;
+		});
+
+		this.scaleListener = projectile.on('scale', (data: EntityScale) => {
```
Comment (pavle-goloskokovic) [inline ts/src/renderer/phaser/classes/PhaserProjectile.ts:34]: please do these types inline and delete d.ts files

**Review by pavle-goloskokovic — APPROVED** [review]
